clone/internal/download/batch.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// download contains classes for downloading data from logs.
+package download
+
+import (
+	"fmt"
+	"time"
+
+	backoff "github.com/cenkalti/backoff/v4"
+	"github.com/golang/glog"
+)
+
+// BatchFetch should be implemented to provide a mechanism to fetch a range of leaves.
+// Enough leaves should be fetched to fully fill `leaves`, or an error should be returned.
+type BatchFetch func(start uint64, leaves [][]byte) error
+
+// Bulk keeps downloading leaves starting from `first`, using the given leaf fetcher.
+// The number of workers and the batch size to use for each of the fetch requests are also specified.
+// The resulting leaves are returned in order over `leafChan`, and any terminal errors are returned via `errc`.
+// Internally this uses exponential backoff on the workers to download as fast as possible, but no faster.
+func Bulk(first uint64, batchFetch BatchFetch, workers, batchSize uint, leafChan chan<- []byte, errc chan<- error) {
```

Comment:
TODO?: probably should take a `context.Context` here to be able to gracefully exit

Reply:
Added TODO. Very timely, as I had this exact deliberation with the verification PR built on top of this and decided I'd do them consistently "later".